BIO_dgram support for BIO_sendmmsg/BIO_recvmmsg

hlandau opened the pull request:
Flags like `BIO_DONTWAIT` ended up getting dropped because Windows can't support it.

There is a new API which must be used to enable local address specification/detection explicitly.

Some previous discussion in https://github.com/openssl/openssl/pull/18238.

hlandau commented:
The errno packing thing is a bit goofy. Could just add an `int *err` argument.

WSASendMsg requires Vista so needs `_WIN32_WINNT` to be bumped to 0x0600. Could probe for it dynamically with GetProcAddress and use it when available. Actually this would help mingw compatibility since it's missing headers for this.

hlandau commented:
It turns out WSASendMsg and WSARecvMsg, which were introduced in Vista, are not exported as normal symbols by Windows. Instead, you have to use them by obtaining function pointers to them, by creating a socket and calling ioctl on it. *What*.

In any case, this does have the advantage that we probe for this functionality at runtime and will fall back on platforms without it (XP, etc.).

When a datagram is sent from the local machine, Windows doesn't provide source/destination information in the control message received from WSARecvMsg even if IP_PKTINFO is enabled. Instead, Control.len is just set to 0 when WSARecvMsg returns. The API as documented in BIO_sendmmsg(3) has been updated to reflect this OS-specific quirk. In this case we zero the output BIO_ADDR.

hlandau commented:
@mattcaswell I think this is ready for review now.

t8m commented on crypto/bio/bss_dgram.c:
```diff
@@ -808,6 +989,620 @@ static int dgram_puts(BIO *bp, const char *str)
+    iov->iov_len  = msg->data_len;
+
+    /* macOS requires msg_namelen be 0 if msg_name is NULL */
+    mh->msg_name        = msg->peer != NULL ? &msg->peer->sa : NULL;
```
Similarly here.

t8m commented on crypto/bio/bss_dgram.c:
```diff
@@ -394,6 +501,44 @@ static long dgram_get_mtu_overhead(bio_dgram_data *data)
+#if OPENSSL_USE_IPV6
+    if (af == AF_INET6) {
+# if defined(IPV6_RECVPKTINFO)
+        if (setsockopt(b->num, IPPROTO_IPV6, IPV6_RECVPKTINFO, &enable, sizeof(enable)) < 0)
```
And here.

t8m commented on crypto/bio/bss_dgram.c:
```diff
@@ -808,6 +989,620 @@ static int dgram_puts(BIO *bp, const char *str)
+    CMSGHDR_TYPE *cmsg;
+    int af = dgram_get_sock_family(b);
+
+    for (cmsg=BIO_CMSG_FIRSTHDR(mh); cmsg != NULL; cmsg=BIO_CMSG_NXTHDR(mh, cmsg)) {
```
Add spaces around `=` and split the line?

t8m commented on crypto/bio/bss_dgram.c:
```diff
@@ -808,6 +989,620 @@ static int dgram_puts(BIO *bp, const char *str)
+         * or the one we are bound to. (Better to error than silently
+         * ignore this.)
+         */
+        if (   local->s_in.sin_port != 0
```
Please drop the extra spaces after `(`

t8m commented on crypto/bio/bss_dgram.c:
```diff
@@ -808,6 +989,620 @@ static int dgram_puts(BIO *bp, const char *str)
+        *info = local->s_in.sin_addr;
+
+        /* See comment above. */
+        if (   local->s_in.sin_port != 0
```
Same as above

t8m commented on crypto/bio/bss_dgram.c:
```diff
@@ -808,6 +989,620 @@ static int dgram_puts(BIO *bp, const char *str)
+         * See comment above, but also applies to the other fields
+         * in sockaddr_in6.
+         */
+        if (   local->s_in6.sin6_port != 0
+            && data->local_addr.s_in6.sin6_port != local->s_in6.sin6_port)
+            return 0;
+
+        if (   local->s_in6.sin6_scope_id != 0
```
And same here

t8m commented on crypto/bio/bss_dgram.c:
```diff
@@ -808,6 +989,620 @@ static int dgram_puts(BIO *bp, const char *str)
+}
+#endif
+
+static ossl_ssize_t dgram_sendmmsg(BIO *b, BIO_MSG *msg, size_t stride, size_t num_msg, uint64_t flags) {
```
Line split, please

t8m commented on crypto/bio/bss_dgram.c:
```diff
@@ -808,6 +989,620 @@ static int dgram_puts(BIO *bp, const char *str)
+#endif
+
+#if M_METHOD == M_METHOD_RECVMMSG
+    /*
+     * If MSG_WAITFORONE is defined, take this as a cue that recvmmsg is
+     * available. The messages passed to us are chunked into one or more
+     * recvmmsg calls. We have to do this because we need to allocate our
+     * translated struct msghdr and struct iovec on the stack to support
+     * multithreaded use.
+     */
```
This comment seems misplaced/spurious?

t8m commented on crypto/bio/bss_dgram.c:
```diff
@@ -808,6 +989,620 @@ static int dgram_puts(BIO *bp, const char *str)
+        jlim = BIO_MAX_MSGS_PER_CALL;
+        if (jlim > num_msg-i)
+            jlim = num_msg-i;
+        for (j=0; j<jlim; ++j) {
```
Same here

t8m commented on crypto/bio/bss_dgram.c:
```diff
@@ -808,6 +989,620 @@ static int dgram_puts(BIO *bp, const char *str)
+                     * If we have done at least one message, we must return the
+                     * count; if we haven't done any, we can give an error code
+                     */
+                    return num_done > 0 ? num_done : -3;
+
+                if (pack_local(b, &mh[j].msg_hdr, BIO_MSG_N(msg, i+j).local) < 1)
+                    return num_done > 0 ? num_done : -3;
```
Why the -3 special return code? If we really need to distinguish it, it should be a defined constant but I do not really see the need. I.E., why not -1?

t8m commented on crypto/bio/bss_dgram.c:
```diff
@@ -808,6 +989,620 @@ static int dgram_puts(BIO *bp, const char *str)
+        /* Do the batch */
+        ret = sendmmsg(b->num, mh, jlim, sysflags);
+        if (ret > 0) {
+            for (j=0; j<(size_t)ret; ++j) {
```
Spaces

t8m commented on crypto/bio/bss_dgram.c:
```diff
@@ -808,6 +989,620 @@ static int dgram_puts(BIO *bp, const char *str)
+    /*
+     * If recvmsg is available, we emulate recvmmsg using multiple calls.
+     */
+    for (i=0; i<num_msg; ++i) {
```
Spaces

t8m commented on crypto/bio/bss_dgram.c:
```diff
@@ -808,6 +989,620 @@ static int dgram_puts(BIO *bp, const char *str)
+
+#elif M_METHOD == M_METHOD_RECVMSG
+    /*
+     * If recvmsg is available, we emulate recvmmsg using multiple calls.
```
sendmsg?

t8m commented on crypto/bio/bss_dgram.c:
```diff
@@ -808,6 +989,620 @@ static int dgram_puts(BIO *bp, const char *str)
+        if (jlim > num_msg-i)
+            jlim = num_msg-i;
+        for (j=0; j<jlim; ++j) {
+            translate_msg(&mh[j].msg_hdr, &iov[j], control[j], &BIO_MSG_N(msg, i+j));
```
Split the line? and spaces around `+`

t8m commented on crypto/bio/bss_dgram.c:
```diff
@@ -808,6 +989,620 @@ static int dgram_puts(BIO *bp, const char *str)
+                 * *(msg->peer) will have been filled in by recvmmsg;
+                 * for msg->local we parse the control data returned
+                 */
+                if (BIO_MSG_N(msg, i+j).local != NULL)
+                    if (extract_local(b, &mh[j].msg_hdr, BIO_MSG_N(msg, i+j).local) < 1)
```
spaces around `+`, split the long line

t8m commented on crypto/bio/bss_dgram.c:
```diff
@@ -808,6 +989,620 @@ static int dgram_puts(BIO *bp, const char *str)
+     * If CMSG_LEN is defined, take this as a cue that recvmsg is available.
+     * We emulate recvmmsg using multiple calls.
+     */
+    for (i=0; i<num_msg; ++i) {
```
spaces

t8m commented on doc/man3/BIO_sendmmsg.pod:
```diff
@@ -0,0 +1,160 @@
+
+If an error occurs before any message has been sent or received, a negative
+value is returned. -2 is returned if the BIO_sendmmsg() or BIO_recvmmsg() method
+is not supported by the BIO; -3 is returned if a local has been specified or
```
I do not think the -3 return is a good idea, I'd just use -1. You have the `BIO_dgram_get_local_addr_cap()` to properly check.

t8m commented on doc/man3/BIO_sendmmsg.pod:
```diff
@@ -0,0 +1,160 @@
+by the return value, which (if nonnegative) expresses the number of messages
+processed.
+
+The caller should set the B<data> member of a B<BIO_MSG> to a buffer containing
+the data to send, or to be filled with a received message. B<data_len> should be
```
I think these should be `I<data>` and `I<data_len>`

t8m commented on doc/man3/BIO_sendmmsg.pod:
```diff
@@ -0,0 +1,160 @@
+that B<BIO_MSG>'s array index), B<data_len> will be modified to specify the
+actual amount of data sent or received.
+
+The B<flags> field of a B<BIO_MSG> provides input per-message flags to the
```
`I<flags>`

t8m commented on doc/man3/BIO_sendmmsg.pod:
```diff
@@ -0,0 +1,160 @@
+flags which affect the entire invocation. No global flags are currently
+defined and this argument must be set to zero.
+
+When these functions are used to send and receive datagrams, the B<peer> field
```
`I<peer>`

t8m commented on doc/man3/BIO_sendmmsg.pod:
```diff
@@ -0,0 +1,160 @@
+datagrams, and written by BIO_recvmmsg() with the source address of received
+datagrams.
+
+Similarly, the B<local> field of a B<BIO_MSG> allows the source address of sent
```
`I<local>`

t8m commented on doc/man3/BIO_sendmmsg.pod:
```diff
@@ -0,0 +1,160 @@
+was the loopback address or the IP address of a local non-loopback interface.
+This is an OS-specific limitation.
+
+The B<stride> argument must be set to B<sizeof(BIO_MSG)>. This argument
```
`I<stride>` `C<sizeof(BIO_MSG)>`

t8m commented on doc/man3/BIO_sendmmsg.pod:
```diff
@@ -0,0 +1,160 @@
+
+=head1 RETURN VALUES
+
+BIO_sendmmsg() and BIO_recvmmsg() the number of messages sent or received. Where
```
`... BIO_recvmmsg() return the...`

t8m commented on doc/man3/BIO_sendmmsg.pod:
```diff
@@ -0,0 +1,160 @@
+
+BIO_sendmmsg() and BIO_recvmmsg() the number of messages sent or received. Where
+a positive value n is returned, all entries in the B<BIO_MSG> array from 0
+through n-1 inclusive have their B<data_len> and B<flags> fields updated with
+the results of the operation on that message. If the call was to recvmmsg() and
+the B<peer> or B<local> fields of that message are non-NULL, the B<BIO_ADDR>
```
`I<>` for parameters or structure members.

t8m commented on doc/man3/BIO_sendmmsg.pod:
```diff
@@ -0,0 +1,160 @@
+successfully sent or received is returned; the caller must learn of the error
+via a subsequent call to the same function.
+
+BIO_dgram_set_local_addr_enable() returns 1 if local address support was
+successfully enabled or disabled and 0 otherwise.
```
Why do we need explicit `BIO_dgram_set_local_addr_enable()` call? Shouldn't the cap() call be sufficient? I.E., what are the use-cases where we are not interested in the local addr and does it bring enough performance difference to not have it always enabled where possible?

hlandau replied:
We need this because local address support does need to be explicitly enabled by setting socket options. Doing this lazily/automatically would be a problem because we want to support multithreaded use of BIO_sendmmsg/BIO_recvmmsg without locking, so these functions should remain stateless.

t8m commented on test/bio_dgram_test.c:
```diff
@@ -0,0 +1,308 @@
+
+static int test_bio_dgram(void)
+{
+    /* Test without local */
+    if (test_bio_dgram_impl(AF_INET, 0) < 1)
+        return 0;
+
+    if (test_bio_dgram_impl(AF_INET6, 0) < 1)
+        return 0;
+
+    /* Test with local */
+    if (test_bio_dgram_impl(AF_INET, 1) < 1)
+        return 0;
+
+    if (test_bio_dgram_impl(AF_INET6, 1) < 1)
+        return 0;
```
Usually this would be done as multiple testcases test - i.e. switch() by `int idx` param.

t8m commented on test/bio_dgram_test.c:
```diff
@@ -0,0 +1,308 @@
+
+    /* First effort should fail due to missing destination address */
+    ret = BIO_sendmmsg(b1, tx_msg, sizeof(BIO_MSG), 1, 0);
+    if (!TEST_int_le((int)ret, -32))
```
Should you use BIO_IS_ERRNO()?

t8m reviewed (changes requested):
Great work!

kaduk commented on crypto/bio/bss_dgram.c:
```diff
@@ -808,6 +989,620 @@ static int dgram_puts(BIO *bp, const char *str)
+        }
+
+        /*
+         * If we didn't receive as many messages as we asked for,
```
Is this the `sendmmsg` block, where we are sending (not receiving) messages)?

hlandau commented:
Updated.

t8m commented:
ping for second review

mattcaswell commented on crypto/bio/bss_dgram.c:
```diff
@@ -808,6 +992,627 @@ static int dgram_puts(BIO *bp, const char *str)
+    socklen_t slen;
+#endif
+
+    if (num_msg == 0)
```
Should we check if `num_msg > OSSL_SSIZE_MAX`?

mattcaswell commented on crypto/bio/bss_dgram.c:
```diff
@@ -808,6 +992,627 @@ static int dgram_puts(BIO *bp, const char *str)
+#endif
+
+    if (num_msg == 0)
+        return 0;
```
Should there be an error on the stack in this case? Similar comment for other error returns below.

mattcaswell replied:
Ah - ok. So 0 here is actually "successfully sent 0 messages"?

My comment about errors on the stack still applies to other error returns thought.

mattcaswell commented on crypto/bio/bss_dgram.c:
```diff
@@ -808,6 +992,627 @@ static int dgram_puts(BIO *bp, const char *str)
+        return -1;
+
+    slen = sizeof(*msg[0].peer);
+    ret = recvfrom(b->num, msg[0].data,
```
Why do we not simulate `recvmmsg` here by calling this in a loop like we do for `recvmsg`?

mattcaswell commented on crypto/bio/bss_dgram.c:
```diff
@@ -808,6 +992,627 @@ static int dgram_puts(BIO *bp, const char *str)
+    int ret;
+#endif
+#if M_METHOD == M_METHOD_RECVMMSG
+# define BIO_MAX_MSGS_PER_CALL   64
```
Why doesn't this end up being defined twice? Didn't we already define this in dgram_sendmmsg above?

mattcaswell commented on crypto/bio/bss_dgram.c:
```diff
@@ -808,6 +992,627 @@ static int dgram_puts(BIO *bp, const char *str)
+    int sysflags;
+#endif
+
+    if (num_msg == 0)
```
Should this check if `num_msg > OSSL_SSIZE_MAX`?

mattcaswell commented on crypto/bio/bss_dgram.c:
```diff
@@ -808,6 +992,627 @@ static int dgram_puts(BIO *bp, const char *str)
+#endif
+
+    if (num_msg == 0)
+        return 0;
```
Should we add an error on the stack here? Similar comment for other error returns below

hlandau replied:
As I see it this is a no-op...

mattcaswell replied:
Ok - but my comment about other error returns still applies.

mattcaswell commented on crypto/bio/bss_dgram.c:
```diff
@@ -808,6 +992,627 @@ static int dgram_puts(BIO *bp, const char *str)
+         */
+        if (local->s_in.sin_port != 0
+            && data->local_addr.s_in.sin_port != local->s_in.sin_port)
+            return 0;
```
Error on stack?

hlandau commented on include/internal/e_os.h:
```diff
@@ -96,7 +96,7 @@
         * and check for current OS version instead.
         */
-#    define _WIN32_WINNT 0x0501
+#    define _WIN32_WINNT 0x0600
```
My understanding is that  increasing `_WIN32_WINNT` allows us to use newer features, but it does not automatically break compatibility. The availability of WSARecvMsg is probed at runtime and there is a fallback if it is not available.

By bumping this there is a risk we accidentially start depending on APIs not supported on XP. I guess maybe we should support a build with version 0x0501, maybe in CI.

hlandau replied:
My code as stands detects availability of WSARecvMsg-related definitions by whether WSAID_WSARECVMSG is defined. So we could make _WIN32_WINNT configurable and the code would automatically use the best function available, whether it's set to 0x0501 or 0x0600. We could discuss what we want to do about evolving _WIN32_WINNT, though it feels like a separate discussion.

mattcaswell replied:
> we could make _WIN32_WINNT configurable

I think it should already be configurable, i.e. you can pass `-D_WIN32_WINNT=...` at configure time. Does your code require _WIN32_WINNT to be 0x0600 before it will work?

The change here changes the default value for `_WIN32_WINNT` which I don't think we can do without some agreement that it is ok to do so.

t8m replied:
I suppose we should document it in the NOTES.windows then? And possibly also change the default to 0x0600 if OMC agrees so (supposedly the override with 0x0501 should be kept working).

mattcaswell replied:
I raised this at the OMC meeting today, and it led to this proposal:

https://github.com/openssl/general-policies/issues/22

hlandau commented on test/bio_dgram_test.c:
```diff
@@ -0,0 +1,308 @@
+            goto err;
+    }
+
+    testresult = 1;
```
The API isn't guaranteed to send multiple messages in a single call, it is permissible for an implementation to always process a single message and return 1.

mattcaswell replied:
The API allows you to supply multiple messages in one go. It is of course permissible for the implementation to only send 1 - but that doesn't mean we shouldn't test sending multiple messages. In particular I would like to see a test that confirms that calling the API and supplying more than `BIO_MAX_MSGS_PER_CALL` works as expected.

hlandau commented on crypto/bio/bss_dgram.c:
```diff
@@ -808,6 +992,627 @@ static int dgram_puts(BIO *bp, const char *str)
+# endif
+                 sysflags,
+                 msg[0].peer != NULL ? &msg[0].peer->sa : NULL,
+                 msg[0].peer != NULL ? sizeof(*msg[0].peer) : 0);
```
See discussion below

mattcaswell commented:
I'd be ok with the second option. As you say the caller still has to be able to handle the case when less than the number of requested messages is sent anyway so it doesn't add much for the BIO to emulate it.

hlandau commented:
Assorted updates. Now the sendmsg cases etc. send one message. Repeated sendmmsg calls have also been removed.

t8m commented:
CI is relevant

t8m commented on crypto/bio/bss_dgram.c:
```diff
@@ -1274,92 +1288,80 @@ static ossl_ssize_t dgram_sendmmsg(BIO *b, BIO_MSG *msg, size_t stride,
     if (num_msg == 0)
         return 0;
 
+    if (num_msg > OSSL_SSIZE_MAX)
+        num_msg = OSSL_SSIZE_MAX;
```
Now this feels like duplicate with the BIO_MAX_MSGS_PER_CALL below. Should it be just replaced with that check? It won't be wrong even for the sendmsg/sendto implementations.

hlandau replied:
The `BIO_MAX_MSGS_PER_CALL` test is specific to sendmmsg/recvmmsg. `BIO_MAX_MSGS_PER_CALL` is also only defined when using recvmmsg.

It may feel like a duplication, but my feeling here is keeping this separated out by method keeps the code more maintainable.

t8m replied:
OK

t8m commented on crypto/bio/bss_dgram.c:
```diff
@@ -1274,92 +1288,80 @@ static ossl_ssize_t dgram_sendmmsg(BIO *b, BIO_MSG *msg, size_t stride,
-                    return num_done > 0 ? num_done : -1;
-            }
-        }
+    if (num_msg > BIO_MAX_MSGS_PER_CALL)
```
Should the BIO_MAX_MSGS_PER_CALL be a public define so the callers won't needlessly attempt to batch more messages? And document that BIO_MAX_MSGS_PER_CALL is a recommended limit on the BIO_sendmmsg call.

hlandau replied:
I don't think we should make any specific promises about the behaviour of this API in terms of the number of messages it handles at this time. There can be other implementations like BIO_dgram_pair and third party implementations which can do what they want.

Maybe a BIO_ctrl to query a suggested batching factor would  be something worth considering, rather than a constant.

t8m replied:
> Maybe a BIO_ctrl to query a suggested batching factor would be something worth considering, rather than a constant.

Yep, that seems to me as something very useful for callers.

t8m commented on crypto/bio/bss_dgram.c:
```diff
@@ -1450,101 +1452,86 @@ static ossl_ssize_t dgram_recvmmsg(BIO *b, BIO_MSG *msg, size_t stride, size_t n
     if (num_msg == 0)
         return 0;
 
+    if (num_msg > OSSL_SSIZE_MAX)
```
Same comment applies here as above

hlandau commented:
OSX failure is going to be annoying to figure out, but thinking about it.

mattcaswell commented:
I've put an OMC hold on this pending the outcome of the discussion in openssl/general-policies#22. I don't think that hold should stop us from continuing to review the rest of this PR.

hlandau commented:
Updated to fix OS X. This was a pain to debug and my resolution is an educated guess about the OS X kernel.

openssl-machine commented:
This PR is in a state where it requires action by @openssl/omc but the last update was 30 days ago

hlandau commented:
- Rebased on top of #18923
- Change of `_WIN32_WINNT` has been reverted, so this no longer needs an OMC hold
- The functionality it enabled can still be enabled manually by specifying `_WIN32_WINNT` for a build
- How to go about enabling this by default can be discussed subsequently

@matttcaswell I think the OMC hold tag can be removed now, and this is ready for review.

mattcaswell commented:
I've lifted the hold.

> How to go about enabling this by default can be discussed subsequently

I guess that is the discussion in openssl/general-policies#22

mattcaswell commented on crypto/bio/bss_dgram.c:
```diff
@@ -790,6 +978,635 @@ static int dgram_puts(BIO *bp, const char *str)
+             * be necessary otherwise.
+             */
+#if defined(__APPLE__)
+            memset(msg->local, 0, sizeof(*msg->local));
```
Here we call memset, but later (line 1571 below) we call `BIO_ADDR_clear`. This looks odd.

mattcaswell commented on doc/man3/BIO_sendmmsg.pod:
```diff
@@ -0,0 +1,167 @@
+I<local> may not be available on all platforms; on these platforms, these
+functions always fail if I<local> is non-NULL.
+
+On Windows, if I<local> is specified, the B<BIO_ADDR> it points to will be
```
Is this also true in some cases for OSX? At least that's what I understood from the code comments above.

mattcaswell commented on doc/man3/BIO_sendmmsg.pod:
```diff
@@ -0,0 +1,167 @@
+
+Unlike most BIO functions, these functions explicitly support multi-threaded
+use. Multiple concurrent writers and multiple concurrent readers are permitted
+in any combination. As such, these functions do not clear, set, or otherwise
```
We should perhaps add a bit more to this. The guarantee only applies to `BIO_sendmmsg` and `BIO_recvmmsg`. Those functions must not be called at the same time as any other BIO function for the same BIO.

mattcaswell commented on doc/man3/BIO_sendmmsg.pod:
```diff
@@ -0,0 +1,167 @@
+which is also the length of the array pointed to by I<msg>.
+
+Unlike most BIO functions, these functions explicitly support multi-threaded
+use. Multiple concurrent writers and multiple concurrent readers are permitted
```
perhaps modify to: "Multiple concurrent writers and multiple concurrent readers of the same BIO are permitted..."

mattcaswell commented on doc/man3/BIO_sendmmsg.pod:
```diff
@@ -0,0 +1,167 @@
+a packed system error code is returned. The macro BIO_IS_ERRNO() evaluates to 1
+if a BIO_sendmmsg() or BIO_recvmmsg() return value represents a system error
+code. BIO_UNPACK_ERRNO() should then be used to obtain the system error code
+(EWOULDBLOCK, etc.) from the return value.
```
I wonder why we do this. In other BIO calls we just signal error and let the caller examine errno directly.

mattcaswell commented on test/bio_dgram_test.c:
```diff
@@ -0,0 +1,455 @@
+static int compare_addr(const BIO_ADDR *a, const BIO_ADDR *b)
+{
+    struct in_addr xa, xb;
+    struct in6_addr xa6, xb6;
```
Doesn't this need to be enclosed in `defined (OPENSSL_USE_IPV6)`?

mattcaswell commented on test/bio_dgram_test.c:
```diff
@@ -0,0 +1,455 @@
+    }
+#endif
+    else
+        return 0;
```
else branch should use `{}` because the other branches do.

mattcaswell commented on test/bio_dgram_test.c:
```diff
@@ -0,0 +1,455 @@
+    }
+#endif
+    else
+        goto err;
```
Enclose else branch in `{}`

mattcaswell commented on test/bio_dgram_test.c:
```diff
@@ -0,0 +1,455 @@
+    BIO_ADDR *addr1 = NULL, *addr2 = NULL, *addr3 = NULL, *addr4 = NULL,
+             *addr5 = NULL, *addr6 = NULL;
+    struct in_addr ina = {0};
+    struct in6_addr ina6 = {0};
```
Wrap in `defined(OPENSSL_USE_IPV6)`

mattcaswell commented on test/bio_dgram_test.c:
```diff
@@ -0,0 +1,455 @@
+     * Try sending more than can be handled in one sendmmsg call (when using the
+     * sendmmsg implementation)
+     */
+    for (i=0; i<OSSL_NELEM(tx_msg); ++i) {
```
Missing spaces around `=` and `<`

mattcaswell commented on test/bio_dgram_test.c:
```diff
@@ -0,0 +1,455 @@
+     * Try receiving more than can be handled in one recvmmsg call (when using
+     * the recvmmsg implementation)
+     */
+    for (i=0; i<OSSL_NELEM(rx_msg); ++i) {
```
Missing spaces around `=` and `<`

hlandau commented:
Updated and rebased, now that the API is merged. Also addressed (I believe) all outstanding comments.

Unsure what's going on with doc-nits. This PR doesn't change the file it's complaining about.

paulidale commented on crypto/bio/bio_local.h:
```diff
@@ -142,6 +142,12 @@ struct sockaddr *BIO_ADDR_sockaddr_noconst(BIO_ADDR *ap);
 const struct sockaddr *BIO_ADDRINFO_sockaddr(const BIO_ADDRINFO *bai);
+
+# if defined(OPENSSL_SYS_WINDOWS) && defined(WSAID_WSARECVMSG)
+#   define BIO_HAVE_WSAMSG
```
Nit: over indented

paulidale commented on test/bio_dgram_test.c:
```diff
@@ -0,0 +1,463 @@
+    }
+
+    tmplen = slen;
+    if (BIO_ADDR_rawaddress(a, pa, &tmplen) < 1)
```
What's the reasoning for not using the TEST_ macros to generate a useful message?  Their use doesn't preclude the caller also generating a message to further pinpoint the location.

paulidale commented on crypto/bio/bss_dgram.c:
```diff
@@ -774,6 +931,35 @@ static long dgram_ctrl(BIO *b, int cmd, long num, void *ptr)
+
+    case BIO_CTRL_DGRAM_SET_LOCAL_ADDR_ENABLE:
+# if defined(SUPPORT_LOCAL_ADDR)
+        num = (num > 0);
```
nit: spurious parenthesis?

paulidale commented on crypto/bio/bss_dgram.c:
```diff
@@ -790,6 +976,714 @@ static int dgram_puts(BIO *bp, const char *str)
+
+            {
+                bio_dgram_data *data = b->ptr;
+                local->s_in6.sin6_addr     =
```
nit: blank line

paulidale reviewed (approved):
Subject to the few nits being fixed.

I've also asked a question about the test case not indicating a reason for failure but I don't think it's a blocker.

hlandau commented:
Updated.

mattcaswell commented on crypto/bio/bss_dgram.c:
```diff
@@ -790,6 +976,715 @@ static int dgram_puts(BIO *bp, const char *str)
+     * the control buffer even if we aren't actually going to examine the
+     * result.
+     */
+    mh->namelen         = msg->peer != NULL ? (ULONG)sizeof(*msg[0].peer) : 0;
```
I am slightly confused by this. We've gone through an elaborate procedure on lines 987-995 above to set `mh->namelen`. Then we ignore what we came up with and overwrite it here. What am I missing?

hlandau replied:
Hm, good catch.

mattcaswell commented on crypto/bio/bss_dgram.c:
```diff
@@ -790,6 +976,715 @@ static int dgram_puts(BIO *bp, const char *str)
+         */
+        if (local->s_in.sin_port != 0
+            && data->local_addr.s_in.sin_port != local->s_in.sin_port) {
+            ERR_raise(ERR_LIB_BIO, ERR_R_UNSUPPORTED);
```
Perhaps a new error reason code explaining this failure? Same, for the multiple similar instances of the same error below.

hlandau commented:
Updated.

paulidale reviewed (approved):
Still good.

openssl-machine commented:
This pull request is ready to merge

levitte commented on crypto/bio/bss_dgram.c:
```diff
@@ -790,6 +976,714 @@ static int dgram_puts(BIO *bp, const char *str)
+
+        /* See comment above. */
+        if (local->s_in.sin_port != 0
+            && data->local_addr.s_in.sin_port != local->s_in.sin_port) {
```
I just discovered, through a build on FreeBSD, that this is untested code.  The error I got is this:

``` C
crypto/bio/bss_dgram.c:1165:16: error: use of undeclared identifier 'data'
            && data->local_addr.s_in.sin_port != local->s_in.sin_port) {
               ^
```

Through this, I see the following declaration sprinkled into all different variants of code in this function:

``` C
CMSGHDR_TYPE *cmsg;
bio_dgram_data *data = b->ptr;
```

Since those are common for all variants, it would probably be safer if they were declared once and for all at the beginning of this function.

levitte reviewed (commented):
Sorry that this comes so late
